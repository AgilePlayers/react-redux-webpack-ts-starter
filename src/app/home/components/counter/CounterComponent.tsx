import React from 'react';
import './counter.scss';

export interface CounterComponentProps {
  counter: number;
  incrementCounter(value: number): void;
}

class CounterComponent extends React.Component<CounterComponentProps> {
  counterInterval: NodeJS.Timer;

  componentDidMount() {
    const incrementCounter = () => {
      this.props.incrementCounter(this.props.counter + 1);
    };

    this.counterInterval = setInterval(incrementCounter, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.counterInterval);
  }

  render() {
    return <div className="counter">value: {this.props.counter}</div>;
  }
}

export default CounterComponent;
