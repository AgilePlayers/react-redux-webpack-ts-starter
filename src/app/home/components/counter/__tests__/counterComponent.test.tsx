import React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import CounterComponent, { CounterComponentProps } from '../CounterComponent';

describe('home / counter / CounterComponent', () => {
  let component: ShallowWrapper | ReactWrapper;
  let props: CounterComponentProps;

  beforeEach(() => {
    props = {
      counter: 0,
      incrementCounter: jest.fn(),
    };
  });

  test('should render', () => {
    component = shallow(<CounterComponent {...props} />);
    expect(component.length).toEqual(1);
  });

  test('should contains element with div with "counter" class', () => {
    component = shallow(<CounterComponent {...props} />);
    expect(component.find('div.counter').length).toEqual(1);
  });

  test('should rendered element contains "value: 0" text', () => {
    component = shallow(<CounterComponent {...props} />);
    expect(component.text()).toEqual('value: 0');
  });

  test('should componentDidMount be called', () => {
    const componentDidMountSpy = jest.spyOn(
      CounterComponent.prototype,
      'componentDidMount',
    );
    component = mount(<CounterComponent {...props} />);
    expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
    componentDidMountSpy.mockClear();
    jest.clearAllTimers();
  });

  test('should setInterval in componentDidMount be called', () => {
    jest.useFakeTimers();
    component = mount(<CounterComponent {...props} />);
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
    jest.clearAllTimers();
  });

  test('should incrementCounter prop function be called after 1000ms', () => {
    jest.useFakeTimers();
    component = mount(<CounterComponent {...props} />);
    expect(component.prop('incrementCounter')).not.toBeCalled();
    jest.advanceTimersByTime(1000);
    expect(component.prop('incrementCounter')).toHaveBeenCalledTimes(1);
    jest.resetAllMocks();
    jest.clearAllTimers();
  });

  test('should incrementCounter prop function stop calling after component unmounting', () => {
    jest.useFakeTimers();
    component = mount(<CounterComponent {...props} />);
    component.unmount();
    expect(clearInterval).toHaveBeenCalledTimes(1);
    jest.resetAllMocks();
    jest.clearAllTimers();
  });

  test('should match snapshot', () => {
    component = shallow(<CounterComponent {...props} />);
    expect(component).toMatchSnapshot();
  });
});
