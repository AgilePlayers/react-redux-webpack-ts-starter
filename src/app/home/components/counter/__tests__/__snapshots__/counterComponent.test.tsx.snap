// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home / counter / CounterComponent should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CounterComponent
    counter={0}
    incrementCounter={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "value: ",
        0,
      ],
      "className": "counter",
    },
    "ref": null,
    "rendered": Array [
      "value: ",
      0,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "value: ",
          0,
        ],
        "className": "counter",
      },
      "ref": null,
      "rendered": Array [
        "value: ",
        0,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
